src/request/mod.rs

```diff
@@ -0,0 +1,295 @@
+use std::borrow::Cow;
+use std::convert::{From, TryInto};
+use std::io::{prelude::*, BufWriter};
+use std::str;
+
+#[cfg(feature = "compress")]
+use http::header::ACCEPT_ENCODING;
+use http::{
+    header::{HeaderValue, IntoHeaderName, HOST},
+    HeaderMap, Method, StatusCode, Version,
+};
+use url::Url;
+
+use crate::error::{Error, ErrorKind, InvalidResponseKind, Result};
+use crate::parsing::{parse_response, Response};
+use crate::streams::{BaseStream, ConnectInfo};
+
+mod builder;
+mod session;
+mod settings;
+
+pub use builder::RequestBuilder;
+pub use session::Session;
+pub use settings::BaseSettings;
+
+fn header_insert<H, V>(headers: &mut HeaderMap, header: H, value: V) -> Result
+where
+    H: IntoHeaderName,
+    V: TryInto<HeaderValue>,
+    Error: From<V::Error>,
+{
+    let value = value.try_into()?;
+    headers.insert(header, value);
+    Ok(())
+}
+
+fn header_insert_if_missing<H, V>(headers: &mut HeaderMap, header: H, value: V) -> Result
+where
+    H: IntoHeaderName,
+    V: TryInto<HeaderValue>,
+    Error: From<V::Error>,
+{
+    let value = value.try_into()?;
+    headers.entry(header).or_insert(value);
+    Ok(())
+}
+
+fn header_append<H, V>(headers: &mut HeaderMap, header: H, value: V) -> Result
+where
+    H: IntoHeaderName,
+    V: TryInto<HeaderValue>,
+    Error: From<V::Error>,
+{
+    let value = value.try_into()?;
+    headers.append(header, value);
+    Ok(())
+}
+
+/// Represents a request that's ready to be sent. You can inspect this object for information about the request.
+#[derive(Debug)]
+pub struct PreparedRequest<B> {
```

Comment:
If the request builder is now called `RequestBuilder`, this could become `Request`?

Reply:
I didn't change the name of the RequestBuilder, it was already named this way. The PreparedRequest terminology comes from the `requests` library where a request ready to be sent is available in immutable form to inspect before sending.

Reply:
Ah, sorry, seems I got confused here.